### ISSUE TITLE
Question: How best to use pre-commit tools with an IDE?

### ISSUE THREAD

**Issue by sirrus233** [pre-commit/pre-commit#2510]
I've tried to search around and find opinions on this, but a common theme seems to be folks being told they are missing the point of pre-commit. I'd like to avoid this! So here's my question:

At the end of the day, **I want to be able to conveniently use tools managed by pre-commit automatically in my IDE.** Standard stuff like black, flake8, isort, etc. Most IDEs (at least PyCharm and VSCode) want a pointer to the tool binaries in order to do that -- fair enough. So where do I find that? Well...

My understanding is that pre-commit installs its binaries into a shared cache, which may be re-used across projects. It seems like a bad idea to point my IDE to these binaries because my project may update the tool version, but the old cache directory stays behind. In any case, the pseudorandom cache directory names imply to me they're not meant to be particularly discoverable.

So I could setup my IDE to run pre-commit directly, and access the tools through some form of the `pre-commit run` command. That would work, but seems to lock me out of using some useful utilities such as [blackconnect](https://plugins.jetbrains.com/plugin/14321-blackconnect), or anything else that runs tools on *save* instead of at pre-commit time. 

I use poetry in most of my projects, so I could manage dependencies with that and pull the tools from my venv, but that seems suboptimal since the dependency will then be managed twice: once by poetry and once by pre-commit. I could try to sync the dependency version between the two, or setup all my pre-commit hooks with the [system] escape hatch, but I feel you have made it abundantly clear that this is not how pre-commit is intended to be used.

So... what's left? What is your personal setup even? :slightly_smiling_face: 

TL;DR How can I use my pre-commit tools in my IDE such that:
1. The tool version is pinned to the project
2. The tool can be used on-the-fly (e.g. when a file is saved)
3. The tool version is managed by only one system (poetry, pre-commit, pip, other)
4. I am not violating the intent or philosophy of how pre-commit is supposed to be used.

**Comment by asottile** [comment]
my editor calls pre-commit, you could probably make an extension for whatever editor you use to do the same

**Comment by asottile** [comment]
if you search the tracker there's duplicates, please follow the template which prompts you to search

**Comment by sirrus233** [comment]
I searched your issue tracker as best I could prior to filing this issue. I apologize if you feel it wasted your time. Thanks for the response.

**Comment by Berk-Tosun** [comment]
@sirrus233 see https://github.com/floatingpurr/sync_with_poetry
